File: app/src/main/java/ui/fragment/AddEditCategoryFragment.java
package ui.fragment;

import android.os.Bundle;
import android.os.Handler;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

import java.util.Objects;

import hlv.cute.todo.databinding.FragmentAddEditCategoryBinding;
import model.Category;
import viewmodel.AddEditCategoryViewModel;

public class AddEditCategoryFragment extends BaseFragment {

    private static final String CATEGORY_ARGS = "category-args";

    private FragmentAddEditCategoryBinding binding;

    private AddEditCategoryViewModel viewModel;

    private TextInputLayout inpLytName;
    private TextInputEditText edtName;
    private TextView txtTitle;
    private MaterialButton btnAdd;

    public AddEditCategoryFragment() {
    }

    public static AddEditCategoryFragment newInstance(Category category) {
        AddEditCategoryFragment fragment = new AddEditCategoryFragment();
        Bundle args = new Bundle();
        args.putSerializable(CATEGORY_ARGS, category);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        viewModel = new ViewModelProvider(this).get(AddEditCategoryViewModel.class);

        if (getArguments() != null && !getArguments().isEmpty()) {
            Category category = (Category) getArguments().getSerializable(CATEGORY_ARGS);

            if (category != null) {
                viewModel.setEditMode(true);
                viewModel.setCategory(category);
            } else {
                viewModel.setEditMode(false);
            }

        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentAddEditCategoryBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews();
        initLogic();
        handleAction();
    }

    private void initViews() {
        binding.aImgBack.setOnClickListener(view -> back());

        inpLytName = binding.inpName;
        edtName = binding.inpEdtName;
        txtTitle = binding.txtTitle;
        btnAdd = binding.mBtnSave;
    }

    private void initLogic() {
        new Handler().postDelayed(() -> {
            edtName.requestFocus();
            showKeyboard();
        }, 500);

        txtTitle.setText(viewModel.getTitleFragment());
        btnAdd.setText(viewModel.getButtonPrimaryText());

        //for moving cursor to end of line editText
        edtName.setText("");
        edtName.append(viewModel.getEditTextTitle());
    }

    private void handleAction() {
        edtName.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.length() > 0)
                    inpLytName.setError(null);
            }
        });

        btnAdd.setOnClickListener(view -> {
            inpLytName.setError(null);

            if (viewModel.isEditMode()) {
                Category editedCategory = viewModel.editCategory(Objects.requireNonNull(edtName.getText()).toString().trim());

                String res = getCategoryViewModel().validateCategory(editedCategory);

                if (res == null) {
                    getCategoryViewModel().editCategory(editedCategory);
                    getTodoViewModel().fetch(); //need to update todos if category was edited
                    getSearchViewModel().fetch();
                    back();
                    return;
                }

                inpLytName.setError(res);
            } else {
                Category category = viewModel.addCategory(Objects.requireNonNull(edtName.getText()).toString().trim());

                String res = getCategoryViewModel().validateCategory(category);

                if (res == null) {
                    getCategoryViewModel().goToTop();
                    getCategoryViewModel().addCategory(category);
                    back();
                    return;
                }

                inpLytName.setError(res);
            }
        });
    }

}

File: app/src/main/java/ui/fragment/AddEditCategoryFragment.kt
package ui.fragment

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.qualifiers.ActivityContext
import hlv.cute.todo.databinding.FragmentAddEditCategoryBinding
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import model.Category
import ui.component.bindingComponent.BaseViewBindingFragment
import utils.Constants
import utils.KeyboardUtil.focusAndShowKeyboard
import viewmodel.AddEditCategoryViewModel
import javax.inject.Inject

@AndroidEntryPoint
class AddEditCategoryFragment : BaseViewBindingFragment<FragmentAddEditCategoryBinding>() {

    val viewModel by viewModels<AddEditCategoryViewModel>()

    override val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> FragmentAddEditCategoryBinding
        get() = FragmentAddEditCategoryBinding::inflate

    @Inject
    @ActivityContext
    lateinit var iContext: Context

    companion object {
        private const val CATEGORY_ARGS = "category-args"

        @JvmStatic
        fun newInstance(category: Category?): AddEditCategoryFragment {
            val fragment = AddEditCategoryFragment()

            val args = bundleOf(CATEGORY_ARGS to category)
            fragment.arguments = args

            return fragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (arguments != null && requireArguments().isEmpty.not()) {
            val category: Category? = requireArguments().getParcelable(CATEGORY_ARGS)

            if (category != null) {
                viewModel.isEditMode = true
                viewModel.category = category
            } else {
                viewModel.isEditMode = false
            }
        }
    }

    override fun initiate() {
        initLogic()
        handleAction()
    }

    private fun initLogic() {
        binding.aImgBack.setOnClickListener { back() }

        binding.inpEdtName.focusAndShowKeyboard(iContext)

        binding.txtTitle.text = viewModel.getTitleFragment()

        binding.mBtnSave.text = viewModel.getButtonPrimaryText()

        //for moving cursor to end of line editText
        binding.inpEdtName.setText("")
        binding.inpEdtName.append(viewModel.getEditTextTitle())
    }

    private fun handleAction() {
        binding.inpEdtName.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                if (editable.isNotEmpty())
                    binding.inpName.error = null
            }
        })

        binding.mBtnSave.setOnClickListener {
            binding.inpName.error = null

            if (viewModel.isEditMode) {
                val editedCategory =
                    viewModel.editCategory(binding.inpEdtName.text.toString().trim())

                val res = categoryViewModel.validateCategory(editedCategory)

                if (res == null) {
                    categoryViewModel.editCategory(editedCategory)
                    updateTodoDetail()
                    //todoViewModel.fetch() //need to update todos if category was edited
                    //searchViewModel.search()
                    back()

                    return@setOnClickListener
                }

                binding.inpName.error = res

            } else {
                val category = viewModel.addCategory(binding.inpEdtName.text.toString().trim())

                val res = categoryViewModel.validateCategory(category)

                if (res == null) {
                    categoryViewModel.goToTop()
                    categoryViewModel.addCategory(category)
                    back()

                    return@setOnClickListener
                }

                binding.inpName.error = res
            }
        }
    }

    private fun updateTodoDetail() {
        lifecycleScope.launch {
            delay(200L) //need some delay to get new todo from db

            val fragment =
                parentFragmentManager.findFragmentByTag(Constants.FragmentTag.TODO_DETAIL) as TodoDetailFragment?

            fragment?.viewModel?.refreshTodo()
        }
    }

}